cm_modules/train.py

```diff
@@ -0,0 +1,122 @@
+          validation_split=0.1, tensorboard=False, verbose=True, threads=None,
+          **kwds):
+
+    tf.compat.v1.keras.backend.set_session(
+        tf.compat.v1.Session(
+            config=tf.compat.v1.ConfigProto(
+                intra_op_parallelism_threads=threads,
+                inter_op_parallelism_threads=threads,
+            )
+        )
+    )
+    model = network.model
+    loss = network.loss
+
+    if output_dir is not None:
+        os.makedirs(output_dir, exist_ok=True)
+
+    if learning_rate is None:
+        optimizer = opt.__dict__[optimizer](clipvalue=clip_grad)
+    else:
+        optimizer = opt.__dict__[optimizer](lr=learning_rate, clipvalue=clip_grad)
+    print("about to compiile")
+
+    # Instantiate an optimizer.
+    # if optimizer == "Adam":
+    #     optimizer = tf.keras.optimizers.Adam()
+
+    print("optimizer", optimizer)
+    model.compile(loss=loss, optimizer=optimizer, experimental_run_tf_function=False)
+
+    # Callbacks
+    callbacks = []
+
+    if save_weights and output_dir is not None:
+        checkpointer = ModelCheckpoint(filepath="%s/weights.hdf5" % output_dir,
+                                       verbose=verbose,
+                                       save_weights_only=True,
+                                       save_best_only=True)
+        callbacks.append(checkpointer)
+    if reduce_lr:
+        lr_cb = ReduceLROnPlateau(monitor='val_loss', patience=reduce_lr, verbose=verbose)
+        callbacks.append(lr_cb)
+    if early_stop:
+        es_cb = EarlyStopping(monitor='val_loss', patience=early_stop, verbose=verbose)
+        callbacks.append(es_cb)
+    if tensorboard:
+        tb_log_dir = os.path.join(output_dir, 'tb')
+        tb_cb = TensorBoard(log_dir=tb_log_dir, histogram_freq=1, write_grads=True)
+        callbacks.append(tb_cb)
+
+    if verbose:
+        model.summary()
+
+    # The format of the input data is modified to use pandas dataframe instead
+    # of scanpy object because getting the following error:
+    #
+    inputs = {'count': adata.X, 'size_factors': adata.obs.size_factors}
+    print("inputs", inputs)
+
+    # Make size_factors a vector of ones
+    # size_factors = np.ones(adata.shape[0])
+    # inputs = {'count': adata, 'size_factors': size_factors}
+
+    if output_subset:
+        gene_idx = [np.where(adata.raw.var_names == x)[0][0] for x in output_subset]
+        output = adata.raw.X[:, gene_idx] if use_raw_as_output else adata.X[:, gene_idx]
+    else:
+        output = adata.raw.X if use_raw_as_output else adata.X
+        # output = np.array(adata.iloc[1:,1:])
+    print("output", output)
+    loss = model.fit(inputs, output,
+                     epochs=epochs,
+                     batch_size=batch_size,
+                     shuffle=True,
+                     callbacks=callbacks,
+                     validation_split=validation_split,
+                     verbose=verbose,
+                     **kwds)
```

Comment:
Beyond what I've said so far I have no idea what is causing the error. It seems like the optimizer is getting deleted or recreated after the first epoch, but I don't know what would cause that. I also don't know how to debug things in tensorflow because none of the training logic is accessible